File: python/src/totalreclaw/imports/adapters/gemini_adapter.py
"""
Gemini import adapter -- thin shim over the shared Rust core parser.

All Gemini format parsing (MyActivity.json, legacy Takeout HTML, and pasted
"Saved info") lives in ``totalreclaw_core.parse_gemini`` so the logic --
including the locale-robust, lossless timestamp handling -- is identical across
every client (Python/Hermes via PyO3, the TypeScript clients via WASM).

This adapter owns only file I/O and the 500MB/RAM preflight; it then delegates
parsing to core. ``totalreclaw-core`` is a hard dependency of this package (it
also backs crypto), so there is no Python-side parser fallback.
"""

import json
import math
import os
from typing import Optional, Callable, List

import psutil

from .base_adapter import BaseImportAdapter
from .types import AdapterParseResult, ConversationChunk, ParsedTurn


class GeminiAdapter(BaseImportAdapter):
    source = 'gemini'
    display_name = 'Google Gemini'

    def parse(
        self,
        *,
        content: Optional[str] = None,
        file_path: Optional[str] = None,
        on_progress: Optional[Callable] = None,
    ) -> AdapterParseResult:
        warnings: List[str] = []
        errors: List[str] = []

        if content:
            raw = content
        elif file_path:
            raw, preflight_err = self._read_file(file_path, warnings, errors)
            if preflight_err is not None:
                return preflight_err
        else:
            errors.append(
                'Gemini import requires either content or file_path. '
                'Export from Google Takeout (takeout.google.com -> "My Activity" -> '
                '"Gemini Apps"); provide the "My Activity.html" (or MyActivity.json) '
                'file path, or paste your Saved info.',
            )
            return AdapterParseResult(
                facts=[], chunks=[], total_messages=0,
                warnings=warnings, errors=errors,
            )

        if on_progress:
            on_progress({'current': 0, 'total': 0, 'phase': 'parsing',
                         'message': 'Parsing Gemini export...'})

        try:
            import totalreclaw_core
            data = json.loads(totalreclaw_core.parse_gemini(raw))
        except Exception as e:
            errors.append(
                'Gemini parsing requires totalreclaw-core >= 2.5.0 (the shared native '
                f'parser); it could not be loaded: {e}',
            )
            return AdapterParseResult(
                facts=[], chunks=[], total_messages=0,
                warnings=warnings, errors=errors,
            )

        if on_progress:
            on_progress({
                'current': len(data.get('chunks', [])),
                'total': len(data.get('chunks', [])),
                'phase': 'parsing',
                'message': (
                    f"Parsed {data.get('total_messages', 0)} messages into "
                    f"{len(data.get('chunks', []))} chunks"
                ),
            })

        return self._result_from_core(data, warnings, errors)

    # ── File read + preflight (the only client-native responsibility) ────────

    def _read_file(self, file_path, warnings, errors):
        """Read a file with the 500MB / RAM preflight. Returns (text, error_result).

        On success returns (text, None); on failure returns ('', AdapterParseResult).
        """
        try:
            resolved = os.path.expanduser(file_path)
            stat = os.stat(resolved)
            file_size_mb = stat.st_size / (1024 * 1024)
            if file_size_mb > 500:
                errors.append(
                    f'File is too large to import: {file_size_mb:.1f}MB exceeds the 500MB cap. '
                    'Split the file into smaller chunks and import each separately.',
                )
                return '', AdapterParseResult(
                    facts=[], chunks=[], total_messages=0,
                    warnings=warnings, errors=errors,
                )
            free_mem = psutil.virtual_memory().available
            if free_mem < stat.st_size * 2:
                free_mb = free_mem / (1024 * 1024)
                need_mb = math.ceil(stat.st_size * 2 / (1024 * 1024))
                errors.append(
                    f'Not enough free memory: {free_mb:.0f}MB available, '
                    f'~{need_mb}MB needed (2× file size). '
                    'Close other applications or split the file.',
                )
                return '', AdapterParseResult(
                    facts=[], chunks=[], total_messages=0,
                    warnings=warnings, errors=errors,
                )
            with open(resolved, 'r', encoding='utf-8') as f:
                return f.read(), None
        except Exception as e:
            errors.append(f'Failed to read file: {e}')
            return '', AdapterParseResult(
                facts=[], chunks=[], total_messages=0,
                warnings=warnings, errors=errors,
            )

    # ── Convert core ParseResult -> AdapterParseResult ───────────────────────

    @staticmethod
    def _result_from_core(data: dict, warnings: List[str], errors: List[str]) -> AdapterParseResult:
        chunks = [
            ConversationChunk(
                title=c.get('title', 'Gemini session'),
                messages=c.get('messages', []),
                timestamp=c.get('timestamp'),
            )
            for c in data.get('chunks', [])
        ]
        # #368 Part 2 — flat per-turn view with real per-turn timestamps. Absent
        # on older core wheels (pre-Part-2) and for the timestamp-less "Saved
        # info" paste format; the import engine falls back to the chunk-level
        # approximation when this list is empty.
        turns = [
            ParsedTurn(
                user_text=t.get('user_text', ''),
                assistant_text=t.get('assistant_text', ''),
                text=t.get('text', ''),
                chunk_index=t.get('chunk_index', 0),
                ts_iso=t.get('ts_iso'),
                ts_unix=t.get('ts_unix'),
                # Range fields present only on core wheels with #368 Part 2
                # straddle fidelity; None on older wheels (engine then assigns
                # whole chunks without straddle-splitting).
                chunk_msg_start=t.get('chunk_msg_start'),
                chunk_msg_end=t.get('chunk_msg_end'),
            )
            for t in data.get('turns', [])
        ]
        meta = {
            'format': data.get('format'),
            'chunks_count': len(chunks),
            'total_messages': data.get('total_messages', 0),
        }
        if data.get('records_count'):
            meta['records_count'] = data['records_count']
        if data.get('skipped'):
            meta['skipped_non_gemini'] = data['skipped']
        return AdapterParseResult(
            facts=[],
            chunks=chunks,
            total_messages=data.get('total_messages', 0),
            warnings=warnings + list(data.get('warnings', [])),
            errors=errors + list(data.get('errors', [])),
            source_metadata=meta,
            turns=turns,
        )

File: skill/plugin/import-adapters/gemini-adapter.ts
import { BaseImportAdapter } from './base-adapter.js';
import type {
  ImportSource,
  AdapterParseResult,
  ConversationChunk,
  ProgressCallback,
} from './types.js';
import fs from 'node:fs';
import os from 'node:os';
import { createRequire } from 'node:module';

// All Gemini format parsing (MyActivity.json, legacy HTML, Saved-info paste)
// lives in the shared Rust core (`@totalreclaw/core` WASM) so the logic —
// including the locale-robust, lossless timestamp handling — is identical across
// every client (Python/Hermes via PyO3, TS via WASM). This adapter is a thin
// shim: it owns only file I/O + the size/RAM preflight, then delegates parsing.
const requireWasm = createRequire(import.meta.url);
let _wasm: typeof import('@totalreclaw/core') | null = null;
function getWasm() {
  if (!_wasm) _wasm = requireWasm('@totalreclaw/core');
  return _wasm!;
}

/** Shape returned by core `parseGemini` (serde -> JS object, snake_case). */
interface CoreParseResult {
  chunks: ConversationChunk[];
  total_messages: number;
  warnings: string[];
  errors: string[];
  format: string;
  records_count?: number;
  skipped?: number;
}

export class GeminiAdapter extends BaseImportAdapter {
  readonly source: ImportSource = 'gemini';
  readonly displayName = 'Google Gemini';

  async parse(
    input: { content?: string; file_path?: string },
    onProgress?: ProgressCallback,
  ): Promise<AdapterParseResult> {
    const warnings: string[] = [];
    const errors: string[] = [];

    let content: string;

    if (input.content) {
      content = input.content;
    } else if (input.file_path) {
      try {
        const resolved = input.file_path.replace(/^~/, os.homedir());
        const fileStat = fs.statSync(resolved);
        const fileSizeMB = fileStat.size / (1024 * 1024);
        if (fileSizeMB > 500) {
          errors.push(
            `File is too large to import: ${fileSizeMB.toFixed(1)}MB exceeds the 500MB cap. ` +
            'Split the file into smaller chunks and import each separately.',
          );
          return { facts: [], chunks: [], totalMessages: 0, warnings, errors };
        }
        const freeMem = os.freemem();
        if (freeMem < fileStat.size * 2) {
          errors.push(
            `Not enough free memory: ${(freeMem / (1024 * 1024)).toFixed(0)}MB available, ` +
            `~${Math.ceil(fileStat.size * 2 / (1024 * 1024))}MB needed (2× file size). ` +
            'Close other applications or split the file.',
          );
          return { facts: [], chunks: [], totalMessages: 0, warnings, errors };
        }
        content = fs.readFileSync(resolved, 'utf-8');
      } catch (e) {
        errors.push(`Failed to read file: ${e instanceof Error ? e.message : 'Unknown error'}`);
        return { facts: [], chunks: [], totalMessages: 0, warnings, errors };
      }
    } else {
      errors.push(
        'Gemini import requires either content or file_path. ' +
        'Export from Google Takeout: takeout.google.com → "My Activity" → "Gemini Apps". ' +
        'Provide the "My Activity.html" (or MyActivity.json) file path, or paste your Saved info.',
      );
      return { facts: [], chunks: [], totalMessages: 0, warnings, errors };
    }

    if (onProgress) {
      onProgress({ current: 0, total: 0, phase: 'parsing', message: 'Parsing Gemini export...' });
    }

    // Delegate ALL format parsing to the shared core.
    const parseResult = getWasm().parseGemini(content) as CoreParseResult;

    if (onProgress) {
      onProgress({
        current: parseResult.chunks.length,
        total: parseResult.chunks.length,
        phase: 'parsing',
        message: `Parsed ${parseResult.total_messages} messages into ${parseResult.chunks.length} chunks`,
      });
    }

    return {
      facts: [],
      chunks: parseResult.chunks,
      totalMessages: parseResult.total_messages,
      warnings: [...warnings, ...parseResult.warnings],
      errors: [...errors, ...parseResult.errors],
      source_metadata: {
        format: parseResult.format,
        chunks_count: parseResult.chunks.length,
        total_messages: parseResult.total_messages,
        ...(parseResult.records_count ? { records_count: parseResult.records_count } : {}),
        ...(parseResult.skipped ? { skipped_non_gemini: parseResult.skipped } : {}),
      },
    };
  }
}
